tests/test_flow/test_persistence_random.py

```diff
@@ -0,0 +1,375 @@
+
+    def add_random_binding(self):
+        dep_entity_names = self._random_dep_entity_names()
+        for dep_entity_name in dep_entity_names:
+            self._child_counts_by_entity_name[dep_entity_name] += 1
+        should_persist = self._random_bool(p_true=self._config.p_persist)
+        descriptor = self._random_descriptor_of_new_entities(should_persist)
+        self._harness.add_binding(descriptor, dep_entity_names)
+
+    def update_random_entity(self):
+        entity_name = self._random_existing_entity_name()
+        self._harness.update_binding(entity_name)
+
+    def query_and_check_random_entity(self):
+        entity_name = self._random_existing_entity_name()
+        self._harness.query_and_check_entity(entity_name)
+
+    def _random_bool(self, p_true=0.5):
+        assert 0.0 <= p_true <= 1.0
+        return self._rng.random() < p_true
+
+    def _random_dep_entity_names(self):
+        # We'll generate a random list of dependencies for a new binding, with the
+        # goal of creating realistic flow graphs that aren't too expensive to compute.
+
+        # If no entities exist yet, our list will have to be empty.
+        n_existing_entities = len(self._child_counts_by_entity_name)
+        if n_existing_entities == 0:
+            return []
+
+        # We don't want the number of top-level entities (those with no dependencies)
+        # to grow linearly with the size of the graph, so we'll decrease the
+        # probability of these as we go.
+        if self._random_bool(p_true=1.0 / n_existing_entities):
+            return []
+
+        # We want to favor deep graphs where each node has similar numbers of parents
+        # and children, regardless of whether it's added early or late. So, we choose
+        # the number of parents (dependencies) with an expected value of 2, and we favor
+        # parents that don't have as many children yet.
+        entity_names = list(self._child_counts_by_entity_name.keys())
+        child_counts = np.array(list(self._child_counts_by_entity_name.values()))
+        raw_weights = np.exp2(-child_counts)
+        weights = raw_weights / raw_weights.sum()
+
+        # We use the geometric distribution, which will usually be between 1 and 3, but
+        # can sometimes give arbitrarily large values.
+        expected_n_samples = 2
+        n_samples = min(len(entity_names), self._rng.geometric(1 / expected_n_samples))
+
+        return list(
+            self._rng.choice(entity_names, size=n_samples, p=weights, replace=False)
+        )
+
+    def _random_existing_entity_name(self):
+        return self._rng.choice(self._harness.get_all_entity_names())
+
+    def _random_descriptor_of_new_entities(self, should_persist):
+        def new():
+            entity_name = self._harness.create_entity(should_persist)
+            self._child_counts_by_entity_name[entity_name] = 0
+            return entity_name
+
+        if self._random_bool(p_true=1 - self._config.p_multi_out):
```

Comment:
super nit: can you invert this if condition? Then, you won't have to complement the multi-output probability.

Reply:
Yep, I agree this would improve it. I actually have a follow-up PR (adding support for tuple descriptors) that fixes this, so I may wait for that one if you don't mind.